### DISCUSSION TITLE
Docs/Wiki, explain the aspect of hardware acceleration in conjunction with FreeRDP

### DISCUSSION THREAD

**Discussion by lorn10** [FreeRDP/FreeRDP#10435] [General]
Hi all

Here follows a noob question about the overall situation of hardware acceleration support in FreeRDP. I have not really an idea how this works at the moment. Maybe some sort of clarification could be added in the Wiki.

On Windows 7 (SP1 and later) there was introduced a long time ago a functionality called [RemoteFX](https://github.com/FreeRDP/FreeRDP/wiki/RemoteFX). Simply stated, this allowed an RDP client to use (to some degree) the GPU of the host RDP server.

The minimal requirement was back then a **Direct3D 9** capable GPU. This requirement was increased in later RemoteFX / Windows versions and there was additionally also introduced some sort of enhanced virtualization and Hyper-V stuff. Finally the whole RemoteFX functionality was due to security problems deprecated in 2020 and finally removed from all supported Windows versions in 2021.

But I suspect that there exist still some sort of RemoteFX support in FreeRDP. Are there any plans to support some possible alternatives in FreeRDP? 

Regarding the future development. I think as of 2024 FreeRDP still only supports the client side. So FreeRDP is [not capable](https://askubuntu.com/questions/475776/can-freerdp-be-used-as-server) to act as an own RDP server which would be an awesome nice feature. Whatever, for the case that FreeRDP will support at some point a sever mode then it should be possible to implement also a (secure) "RemoteFX 1" based hardware acceleration with the help of the open-source Direct3D 9 API [Gallium Nine](https://github.com/iXit/wine-nine-standalone). And that of course not only for Windows clients but also for Linux ones. :wink:

**Comment by akallabeth** [comment]
you seem to mix quite some stuff together here :)

1. `FreeRDP` as a server here: [ogon](https://github.com/ogon-project/ogon) and [gnome-remote-desktop](https://gitlab.gnome.org/GNOME/gnome-remote-desktop)
2. there is [MS-RDPEGFX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegfx/da5c75f9-cd99-450c-98c4-014a496942b0) which is the replacement and comes with the same codecs (or H264) 
3. hardware support does not have a direct dependency to directx 9 - you just need something to run the encoder/decoder on the graphics card (shader/opencl/...) and have the required features on the card
4. adding our own codecs is kind of useless for most as you need  `FreeRDP` as client & server then

**Comment by lorn10** [comment]
Thanks for the informative answer.

Okay, so it looks that as of 2024 there exist some RDP server support at least for the GNOME environment. Nice to know, - will add that information on askUbuntu. :+1: 

And it is also nice to known that the effective hardware support is not that central in conjunction with RDP. Most likely I was here somehow confused because of those virtualization tools where a client hardware acceleration support via the host hardware can have a big impact. (Like seen here: [virgl/win: Virtio GPU 3D acceleration for windows](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/24223))

Whatever, thanks for the super fast reply. Will close this discussion now.

**Reply by akallabeth** [reply]
@lorn10 client and server side you can accelerate the encoding/decoding.
server side it is more important as you can reduce server resources, while most clients are already fast enough to decode in software.
(just think as streaming your desktop, it basically is just some video codec in between nowadays, RemoteFX (or one of its variants like progressive) or H264 if activated.

**Comment by vikramre1989** [comment]
FastRDP is a tool for brute force attacks on the Remote Desktop Protocol (RDP) on the Windows operating system, which allows users to remotely connect to computers and servers. This tool is written in C# programming language. Also, this tool has a Multi-threading feature that increases the efficiency and speed of testing simultaneous compounds.
I tested this tool, and it works great! I’ll put the link to this tool below.
[FastRDP](https://github.com/rensii-1996/RDP-Brute-Force)

**Comment by MarcDoughty** [comment]
@lorn10 , just to clarify a few things.

There was a RemoteFX vGPU available from the Windows Hyper-V hypervisor. This provided a Direct3D virtual GPU to guest VMs, provided from host GPU resources. Most VMs have software-only video, or you have to do things like pass an entire host GPU to the guest or (with certain types of mostly high-end 'business' GPUs) slice a physical GPU up and serve those slices to guest VMs. This is a host/guest VM thing, providing accelerated video to VMs. Notbly, the RemoteFX vGPU ONLY provided Direct3D, and only a subset of what the physical GPU on the host did. OpenGL 4.6 and Vulkan were not available in it (as far as I recall), but VMs could do Direct3D stuff much faster with it.

There's also RemoteFX features of the RDP protocol. Microsoft is famous for using the same names for different things, and this is one of those examples. Windows 7 had a way to compress RDP that they branded 'RemoteFX', but it didn't have anything to do with the virtual GPU in the first paragraph. It was basically just 'a better way to compress the stream of data going from an RDP server to a client'.

Windows can use GPU acceleration to make an RDP server faster and use less CPU. This is where things get interesting. You could never 'use a server GPU from an RDP client' directly, but if your RDP server has a GPU (whether it's a regular machine with an actual GPU, or a VM getting a 'slice' or a 'passed-through' GPU or that deprecated RemoteFX vGPU), you would have an ' RDP server' where the desktop and apps had the ability to run apps with 3D APIs quickly, AND the actual compression of the RDP data stream was compressed. If you carefully designed your setup with the right stuff, you would experience a remote Windows session that felt much snappier (because the protocol and the windows desktop were both accelerated), and in some configs, you could use advanced 3D APIs that were not available on unaccelerated VMs.

The RemoteFX vGPU was extremely cool mainly because it allowed Windows VMs to feel much snappier AND use less server CPU resources than regular VMs. It was never great if you needed advanced 3D APIs, because the VM only had Direct3D, not OpenGL or Vulkan. This was all very confusing to VDI SYSADMINs, I was in tons of meetings where this stuff went over peoples' heads and various ideas had to be tossed because they wouldn't provide the APIs we needed even though 'regular Windows sessions' ran much faster.

The short version is that if you are connecting to a Windows physical machine via RDP, you likely experience the accelerated experience with apps and the protocol. If you are connecting to a Windows VM, chances are that you are getting a degraded experience, even if the host has a nice GPU. If you have host server hardware that's able to 'slice up' or 'pass through' the GPU on the server to the VM and you set that up (you will know if you do, it's complicated and involves a lot of config), you can have the full experience of acceleration on a VM over an RDP session.

For FreeRDP, and this is a bit out of my wheelhouse, so someone correct me if I'm wrong, the 'server' is typically either a physical or virtual Linux system. If it's physical, then the graphics are accelerated by whatever GPU is on the system, and FreeRDP itself MIGHT have some hooks into using the system's various paths to accelerate the protocol (e.g., using an h.264 library that recognizes things like Quick Sync Video or various GPU-supplied paths to compress frames). If you are connecting to a virtual machine, then it's likely that the graphics on the 'server' are unaccelerated and so is the protocol.

A good way to get an idea if things are accelerated from your RDP server, whether it's Windows or FreeRDP, physical or virtual, is to look at what GPU the system you are connecting to has. If you see something generic or from a hypervisor in Device Manager or 'lspci | grep VGA', things aren't going to be accelerated, but if you see a GPU from Intel, AMD, or NVIDIA, they probably are.

**Reply by lorn10** [reply]
Really many thanks for this super informative answer @MarcDoughty.

The whole matter is effectively far more complex than I thought. And I mixed up various different things most likely because those are described not that clearly / somewhat misleading by Microsoft. :wink: 

Whatever, when it comes to the virtualization aspect, [SoftGPU](https://github.com/JHRobotics/softgpu) has a nice feature which allows the direct pass-trough of D3D9 API functions from the virtual VM to the host GPU as long as [Gallium Nine](https://github.com/iXit/wine-nine-standalone) is present (unfortunately only Mesa 25.1 and older). This gives almost no additional overhead because no D3D to OGL translation is processed.

And this was effectively also the original reason of my question in this topic here. I really thought that FreeRDP could just simply use Gallium Nine (as Windows 7 is using D3D9) and implement its own variant of [RemoteFX](https://github.com/FreeRDP/FreeRDP/wiki/RemoteFX). 

I have now learned that it is not that simple. Because there was not really any real hardware acceleration present with RemoteFX, - at least not in the way I had imagined.
